feanil/api docs (#32971)
openedx/edx-platform: docs/references/auth_code_samples.rst

   {'access_token': 'eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiAibG1zLWtleSIsICJleHAiOiAxNjkyMjExNjM4LCAiZ3JhbnRfdHlwZSI6ICJwYXNzd29yZCIsICJpYXQiOiAxNjkyMjA4MDM4LCAiaXNzIjogImh0dHA6Ly9sb2NhbGhvc3Q6MTgwMDAvb2F1dGgyIiwgInByZWZlcnJlZF91c2VybmFtZSI6ICJmZWFuaWwiLCAic2NvcGVzIjogWyJyZWFkIiwgIndyaXRlIiwgImVtYWlsIiwgInByb2ZpbGUiXSwgInZlcnNpb24iOiAiMS4yLjAiLCAic3ViIjogIjVjMTBmNjZmMmQ2MzkwYjcwNjYyYzkxNGFhZTdlZjc5IiwgImZpbHRlcnMiOiBbInVzZXI6bWUiXSwgImlzX3Jlc3RyaWN0ZWQiOiBmYWxzZSwgImVtYWlsX3ZlcmlmaWVkIjogdHJ1ZSwgImVtYWlsIjogImZlYW5pbEBheGltLm9yZyIsICJuYW1lIjogIkZlYW5pbCBQYXRlbCIsICJmYW1pbHlfbmFtZSI6ICIiLCAiZ2l2ZW5fbmFtZSI6ICIiLCAiYWRtaW5pc3RyYXRvciI6IHRydWUsICJzdXBlcnVzZXIiOiB0cnVlfQ.iGFl7qsAUau0-40oq8Of0f72kguq2Hc_drijCnI2I-M',
    'expires_in': 3600,
    'refresh_token': 'm8iXhVlGABu52xFxVFj5rAz8xSjsRq',
    'scope': 'read write email profile',
    'token_type': 'JWT'}

.. note:: The client type must be ``public`` for this to work.

.. _JWT from application:

Get a JWT with a client_id and client_secret
********************************************

.. code-block::

   import base64
   import requests

   from pprint import pprint

   client_id = "ukbclQB8aPh7hgsy8ifPXkPf7fRqgUq1w21f2YZa"

>    # Note this should actually be secert and probably not in your code but

Reviewer: ```suggestion
   # Note this should actually be secret and probably not in your code but
```

   # provided here in the example
   client_secret = "xkN0BJ19q9Jk8UPUppEtC1xe4764c81ioFtlegvokbmnAC7CFCT5gG1Og5nnFmCNc3NHNhUwWWDRVcBfnLSZ4xAlEmSePzfkFtLE06cwR1MuSc0gx9LUEjRrTs3j2vgK"

   credential = f"{client_id}:{client_secret}"
   encoded_credential = base64.b64encode(credential.encode("utf-8")).decode("utf-8")

> 

Reviewer: Nit: Remove extra line.
```suggestion
```

   {'access_token': 'eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiAibG1zLWtleSIsICJleHAiOiAxNjkyMjExNjM4LCAiZ3JhbnRfdHlwZSI6ICJjbGllbnQtY3JlZGVudGlhbHMiLCAiaWF0IjogMTY5MjIwODAzOCwgImlzcyI6ICJodHRwOi8vbG9jYWxob3N0OjE4MDAwL29hdXRoMiIsICJwcmVmZXJyZWRfdXNlcm5hbWUiOiAiZmVhbmlsIiwgInNjb3BlcyI6IFsicmVhZCIsICJ3cml0ZSIsICJlbWFpbCIsICJwcm9maWxlIl0sICJ2ZXJzaW9uIjogIjEuMi4wIiwgInN1YiI6ICI1YzEwZjY2ZjJkNjM5MGI3MDY2MmM5MTRhYWU3ZWY3OSIsICJmaWx0ZXJzIjogW10sICJpc19yZXN0cmljdGVkIjogZmFsc2UsICJlbWFpbF92ZXJpZmllZCI6IHRydWUsICJlbWFpbCI6ICJmZWFuaWxAYXhpbS5vcmciLCAibmFtZSI6ICJGZWFuaWwgUGF0ZWwiLCAiZmFtaWx5X25hbWUiOiAiIiwgImdpdmVuX25hbWUiOiAiIiwgImFkbWluaXN0cmF0b3IiOiB0cnVlLCAic3VwZXJ1c2VyIjogdHJ1ZX0.CX1S0QGrWKEPOHC8kUzGcvW8Ky04RCA8vU8WJrZURSw',
    'expires_in': 3600,
    'scope': 'read write email profile',
    'token_type': 'JWT'}

.. note:: When you get a JWT using ``client_credentials`` you don't get a
   refresh token.  You're just expected to make a new call with your client
   credentials.

Check to see if a JWT is Expired
********************************

.. code-block::

   import jwt

   # See above examples for how to get a JWT token
   jwt_token = token_request.json()['access_token']

   try:

>        jwt.decode(jwt_token, "secret", audience="lms-key", algorithms=['HS256'])

Reviewer: Not sure if you could add a comment that for Open edX code, we really want to make use of [jwt_decode_handler](https://github.com/openedx/edx-drf-extensions/blob/e8c7e0c743cc1bca6c89953f1cd94badfaaad99c/edx_rest_framework_extensions/auth/jwt/decoder.py#L32), rather than having custom decoding code.

Author: Yea, sounds like we should add a code sample that's `Validate JWT in Open edX Django Services`?  Looking at the comment for `jwt_decode_handler` it doesn't raise an ExpiredSignatureError. Do you know if that's a comment oversight or more complex than that?

Reviewer: It looks like we are missing a test for that, so I don't know what exception it raises. See https://github.com/openedx/edx-drf-extensions/blob/e8c7e0c743cc1bca6c89953f1cd94badfaaad99c/edx_rest_framework_extensions/auth/jwt/tests/test_decoder.py#L45.

In  New Relic in the last month, I only see two instances of `jwt.exceptions:ExpiredSignatureError`, one in discovery stage and one in discovery prod. I _don't think_ that discovery has custom decode login, but that is sprinkled throughout the platform, so I can't guarantee that.

   except jwt.ExpiredSignatureError:
       # Signature has expired

Refresh a JWT Using a Refresh Token
***********************************

.. code-block::

   import requests

>    # See above examples for how to get a JWT token with a refresh token

Reviewer: 1. Doesn't this comment describe this example, and not anything above?
2. How does the user get a refresh token? Is that a missing section?

Author: This comment describes getting a JWT token from a username and password.  Because you can't get a refresh token when using a confidential client as far as I can tell.

Reviewer: The main point is that this example, aptly titled "Refresh a JWT Using a Refresh Token", relies on a `refresh_token`, but the document is missing a section (or note) detailing how I might retrieve such a token.

It seems like this code comment should point me to these instructions.

Author: Fair point, I added a longer comment that's hopefully more clear.

Reviewer: Is the following what you are trying to convey?
```suggestion
   # See "Get a JWT with a Username and Password" for how to get a refresh token
```
If so, it is a little fuzzy because "with" (in your comment) and "using" (in the title of this section) are so similar that I think it was where I was getting confused. Consider my language, or some other update. Thanks.

>    # The reponse from that request will include a `refresh_token` attribute.

Author: ```suggestion
   # The response from that request will include a `refresh_token` attribute.
```

    'expires_in': 3600,
    'scope': 'read write email profile',
    'token_type': 'JWT'}

.. note:: When you get a JWT using ``client_credentials`` you don't get a
   refresh token.  You're just expected to make a new call with your client
   credentials.

Check to see if a JWT is Expired
********************************

.. code-block::

   import jwt

   # See above examples for how to get a JWT token
   jwt_token = token_request.json()['access_token']

   try:
       jwt.decode(jwt_token, "secret", audience="lms-key", algorithms=['HS256'])
   except jwt.ExpiredSignatureError:
       # Signature has expired

Refresh a JWT Using a Refresh Token
***********************************

.. code-block::

   import requests

>    # See above examples for how to get a JWT token with a refresh token
>    # "Get a JWT with a Username and Password"
>    # The reponse from that request will include a `refresh_token`

Author: ```suggestion
   # See "Get a JWT with a Username and Password" for how to get a refresh token.
   # The reponse from that request will include a `refresh_token` attribute.
```
